File: backend/src/main/java/com/example/contenthub/service/crawling/novel/NaverSeriesCrawler.java
package com.example.contenthub.service.crawling.novel;

import com.example.contenthub.constants.crawling.NaverNovelXPath;
import com.example.contenthub.dto.ContentCrawlDTO;
import com.example.contenthub.exception.CrawlerException;
import com.example.contenthub.service.auth.social.NaverSeriesLogin;
import lombok.RequiredArgsConstructor;
import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.TimeoutException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.springframework.stereotype.Component;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Component
@RequiredArgsConstructor
public class NaverSeriesCrawler {
    private final NaverSeriesLogin naverLogin;

    public List<ContentCrawlDTO> crawl(WebDriver mainDriver) {
        List<ContentCrawlDTO> novels = new ArrayList<>();
        try {
            setUpDrivers(mainDriver);
            int totalPages = getLastPageNumber(mainDriver);
            List<String> novelLinks = getNaverSeriesData(mainDriver, totalPages);

            for (String href : novelLinks) {
                try {
                    ContentCrawlDTO content = extractNovelData(href, mainDriver);
                    novels.add(content);
                } catch (TimeoutException e) {
                    System.err.println("TimeoutException 발생 - 다음 항목으로 진행: " + href);
                } catch (Exception e) {
                    System.err.println("예상치 못한 오류 발생 - 다음 항목으로 진행: " + href);
                    //e.printStackTrace();
                }
            }
        } catch (TimeoutException e) {
            throw new CrawlerException(CrawlerException.ExceptionMessage.TIMEOUT_EXCEPTION.getMessage(), e);
        } catch (InterruptedException e) {
            throw new CrawlerException(CrawlerException.ExceptionMessage.UNEXPECTED_EXCEPTION.getMessage(), e);
        }
        return novels;
    }

    private void setUpDrivers(WebDriver mainDriver) {
        mainDriver.get(NaverNovelXPath.URL.getXpath());
        naverLogin.activateBot(mainDriver);
    }

    public int getLastPageNumber(WebDriver driver) {
        try {
            WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
            WebElement pagination = wait.until(
                    ExpectedConditions.presenceOfElementLocated(By.xpath(NaverNovelXPath.PAGE_COUNT.getXpath())));
            List<WebElement> pageLinks = pagination.findElements(By.tagName("a"));
            int totalPages = pageLinks.size();
            return totalPages;
        } catch (Exception e) {
            e.printStackTrace();
            return 1;
        }
    }

    private void navigateToPage(WebDriver driver, String url) {
        driver.get(url);
        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        wait.until(ExpectedConditions.urlToBe(url));
    }

    private List<String> getNaverSeriesData(WebDriver webDriver, int totalPages) throws InterruptedException {
        List<String> novelLinks = new ArrayList<>();
        for (int i = 1; i <= totalPages; i++) {
            String url = NaverNovelXPath.URL.getXpath() + i;
            navigateToPage(webDriver, url);

            WebDriverWait wait = new WebDriverWait(webDriver, Duration.ofSeconds(5));
            List<WebElement> elements = wait.until(driver -> driver.findElements(By.xpath(NaverNovelXPath.LIST.getXpath())));
            for (WebElement el : elements) {
                String href = el.findElement(By.xpath(".//a")).getAttribute("href");
                novelLinks.add(href);
            }
        }
        return novelLinks;
    }

    private ContentCrawlDTO extractNovelData(String detailHref, WebDriver webdriver) {
        String contentId = extractContentId(detailHref);
        String url = NaverNovelXPath.DETAIL_URL.getXpath() + contentId;
        // 상세 페이지 이동
        navigateToPage(webdriver, url);

        WebDriverWait wait = new WebDriverWait(webdriver, Duration.ofSeconds(10));
        wait.until(ExpectedConditions.presenceOfElementLocated(By.xpath(NaverNovelXPath.TITLE.getXpath())));

        // 상세 정보 추출
        String title = extractTitle(webdriver.findElement(By.xpath(NaverNovelXPath.TITLE.getXpath())).getText());
        String description = getDescription(webdriver);
        String coverImg = webdriver.findElement(By.xpath(NaverNovelXPath.COVER_IMG.getXpath())).getAttribute("src");
        boolean isAdultContent = isAdultContent(webdriver);
        String genre = webdriver.findElement(By.xpath(NaverNovelXPath.GENRE.getXpath())).getText();
        return new ContentCrawlDTO(title, description, coverImg, genre, isAdultContent, contentId);
    }

    private boolean isAdultContent(WebDriver driver) {
        return driver.findElements(By.xpath(NaverNovelXPath.IS_ADULT.getXpath())).size() > 0;
    }

    private static String extractTitle(String title) {
        return title.replaceAll("\\[.*?\\]", "").trim();
    }

    private String extractContentId(String input) {
        Pattern pattern = Pattern.compile("\\d+");
        Matcher matcher = pattern.matcher(input);

        if (matcher.find()) {
            return matcher.group();
        } else {
            System.out.println("해당 작품 id를 찾을 수가 없습니다!");
            return null;
        }
    }

    private String getDescription(WebDriver driver) {
        try {

            JavascriptExecutor js = (JavascriptExecutor) driver;
            WebElement element = driver.findElement(By.xpath(NaverNovelXPath.DESCRIPTION_WITH_MORE.getXpath()));
            String des = (String) js.executeScript("return arguments[0].innerText.trim();", element);
            return cleanText(des);
        } catch (NoSuchElementException e) {
            try {
                return driver.findElement(By.xpath(NaverNovelXPath.DESCRIPTION_NO_MORE.getXpath())).getText();
            } catch (NoSuchElementException ex) {
                return "설명 없음";
            }
        }
    }
    private String cleanText(String text) {
        return text.replaceAll("[\\u00A0\\u200B\\u200C\\u200D\\uFEFF]", " ")
                    .replace("접기", "")
                    .trim();
    }

}

File: data_pipeline/services/naver_series_crawler.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException
from selenium.common.exceptions import NoSuchElementException
from time import sleep
import logging
import re
from services.naver_login import activate_bot
from enums.naver_novel_xpath import NaverNovelXPath

class NaverSeriesCrawler:
    def __init__(self, driver: webdriver.Chrome):
        self.driver = driver
        self.driver.get(NaverNovelXPath.URL.value)
        activate_bot(driver)

    def crawl(self):
        novels = []
        try:
            total_pages = self.get_last_page_number()
            novel_links = self.get_naver_series_data(total_pages)
            for href in novel_links:
                try:
                    content = self.extract_novel_data(href)
                    novels.append(content)
                except Exception as e:
                    logging.error(f"예상치 못한 오류 발생 - 다음 항목으로 진행: {href}")
                    # logging.exception(e)
        except TimeoutException as e:
            raise Exception("TimeoutException 발생", e)
        return novels

    def get_last_page_number(self):
        try:
            wait = WebDriverWait(self.driver, 10)
            pagination = wait.until(EC.presence_of_element_located((By.XPATH, NaverNovelXPath.PAGE_COUNT.value)))
            page_links = pagination.find_elements(By.TAG_NAME, "a")
            total_pages = len(page_links)
            return total_pages
        except Exception as e:
            print("에러: {e}")
            return 1

    def navigate_to_page(self, url: str):
        self.driver.get(url)

        wait = WebDriverWait(self.driver, 10)
        wait.until(EC.url_to_be(url))

    def get_naver_series_data(self, total_pages: int) -> list:
        novel_links = []
        for i in range(1, total_pages + 1):

            url = f"{NaverNovelXPath.URL.value}{i}"
            self.navigate_to_page(url)

            wait = WebDriverWait(self.driver, 5)

            elements = wait.until(
                lambda driver: driver.find_elements(By.XPATH, NaverNovelXPath.LIST.value)
            )

            for el in elements:
                href = el.find_element(By.XPATH, ".//a").get_attribute("href")
                novel_links.append(href)

        return novel_links

    def extract_novel_data(self, detail_href: str) -> dict:
        content_id = self.extract_content_id(detail_href)
        url = NaverNovelXPath.DETAIL_URL.build_url(content_id)

        # 상세 페이지로 이동
        self.navigate_to_page(url)

        # 상세 정보 추출
        original_title = self.driver.find_element(By.XPATH, NaverNovelXPath.TITLE.value).text
        title = self.extract_title(original_title)  # 실제 제목 -[] 제외
        is_adult_content = self.contains_adult_tag()  # 성인 여부 체크
        description = self.get_description()
        cover_img = self.driver.find_element(By.XPATH, NaverNovelXPath.COVER_IMG.value).get_attribute("src")
        genre = self.driver.find_element(By.XPATH, NaverNovelXPath.GENRE.value).text
        return {
            "title": title,
            "description": description,
            "cover_img": cover_img,
            "genre": genre,
            "is_adult_content": is_adult_content,
            "content_id": content_id,
        }

    def contains_adult_tag(self):
        return len(self.driver.find_elements(By.XPATH, NaverNovelXPath.IS_ADULT.value)) > 0

    def extract_title(self, title):
        return re.sub(r"\[.*?\]", "", title).strip()

    def extract_content_id(self, input_str):
        match = re.search(r"\d+", input_str)
        if match:
            return match.group()
        else:
            print("해당 작품 id를 찾을 수가 없습니다!")
            return None

    def get_description(self):
        try:
            element = self.driver.find_element(By.XPATH, NaverNovelXPath.DESCRIPTION_WITH_MORE.value)
            description = self.driver.execute_script("return arguments[0].innerText.trim();", element)
            return self.clean_text(description)
        except Exception:
            try:
                description = self.driver.find_element(By.XPATH, NaverNovelXPath.DESCRIPTION_NO_MORE.value).text
                return self.clean_text(description)
            except Exception:
                return "설명 없음"

    def clean_text(self, text):
        # 불필요한 특수문자 및 "접기" 텍스트를 제거
        return text.replace("\u00A0", " ").replace("\u200B", " ").replace("\u200C", " ")\
            .replace("\u200D", " ").replace("\uFEFF", " ").replace("접기", "").strip()
